File: ios/DoerApp/DoerApp/ViewModels/ProjectViewModel.swift
import SwiftUI
import Supabase

@MainActor
final class ProjectViewModel: ObservableObject {
    @Published var projects: [Project] = []
    @Published var sections: [String: [ProjectSection]] = [:] // keyed by projectId
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let projectService = ProjectService()

    // MARK: - Fetch

    func fetchProjects() async {
        isLoading = true
        defer { isLoading = false }
        do {
            projects = try await projectService.fetchAll()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func fetchSections(projectId: String) async {
        do {
            let result = try await projectService.fetchSections(projectId: projectId)
            sections[projectId] = result
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Create

    func createProject(name: String, color: String) async {
        guard let userId = await AuthService().getUser()?.id.uuidString else { return }
        let lastPosition = projects.map(\.position).max()
        let position = PositionHelpers.getPositionAtEnd(lastPosition: lastPosition)

        do {
            let project = try await projectService.create(
                userId: userId, name: name, color: color, position: position
            )
            projects.append(project)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func deleteProject(id: String) async {
        projects.removeAll { $0.id == id }
        do {
            try await projectService.delete(id: id)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Sections

    func createSection(projectId: String, name: String) async {
        guard let userId = await AuthService().getUser()?.id.uuidString else { return }
        let existing = sections[projectId] ?? []
        let lastPosition = existing.map(\.position).max()
        let position = PositionHelpers.getPositionAtEnd(lastPosition: lastPosition)

        do {
            let section = try await projectService.createSection(
                userId: userId, projectId: projectId, name: name, position: position
            )
            sections[projectId, default: []].append(section)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func deleteSection(id: String) async {
        for key in sections.keys {
            sections[key]?.removeAll { $0.id == id }
        }
        do {
            try await projectService.deleteSection(id: id)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Computed

    var inboxProject: Project? {
        projects.first { $0.isInbox }
    }

    var nonInboxProjects: [Project] {
        projects.filter { !$0.isInbox }
    }
}

File: android/DoerApp/app/src/main/java/com/doer/app/ui/projects/ProjectViewModel.kt
package com.doer.app.ui.projects

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.doer.app.data.model.Project
import com.doer.app.data.model.ProjectSection
import com.doer.app.data.repository.ProjectRepository
import com.doer.app.util.PositionHelpers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class ProjectUiState(
    val projects: List<Project> = emptyList(),
    val sections: List<ProjectSection> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null
)

class ProjectViewModel : ViewModel() {

    private val projectRepository = ProjectRepository()

    private val _uiState = MutableStateFlow(ProjectUiState())
    val uiState: StateFlow<ProjectUiState> = _uiState.asStateFlow()

    fun fetchAll(userId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val projects = projectRepository.fetchAll(userId)
                _uiState.update { it.copy(projects = projects, isLoading = false) }
            } catch (e: Exception) {
                _uiState.update { it.copy(isLoading = false, error = e.message) }
            }
        }
    }

    fun createProject(userId: String, name: String, color: String) {
        viewModelScope.launch {
            try {
                val maxPos = _uiState.value.projects.maxOfOrNull { it.position } ?: 0.0
                val position = PositionHelpers.getPositionAtEnd(maxPos)
                projectRepository.create(userId, name, color, position)
                fetchAll(userId)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun updateProject(id: String, name: String, color: String, userId: String) {
        viewModelScope.launch {
            try {
                projectRepository.update(id, name, color)
                fetchAll(userId)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun deleteProject(id: String, userId: String) {
        viewModelScope.launch {
            try {
                projectRepository.delete(id)
                fetchAll(userId)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun fetchSections(projectId: String) {
        viewModelScope.launch {
            try {
                val sections = projectRepository.fetchSections(projectId)
                _uiState.update { it.copy(sections = sections) }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun createSection(projectId: String, userId: String, name: String) {
        viewModelScope.launch {
            try {
                val maxPos = _uiState.value.sections.maxOfOrNull { it.position } ?: 0.0
                val position = PositionHelpers.getPositionAtEnd(maxPos)
                projectRepository.createSection(projectId, userId, name, position)
                fetchSections(projectId)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun deleteSection(sectionId: String, projectId: String) {
        viewModelScope.launch {
            try {
                projectRepository.deleteSection(sectionId)
                fetchSections(projectId)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }

    /**
     * Returns the inbox project for the given user, if one exists.
     */
    fun getInboxProject(): Project? {
        return _uiState.value.projects.find { it.isInbox }
    }
}
